### PR TITLE
c8d/pull: Add CRI-compatible annotation of pulled content

### PR DESCRIPTION
**Pull request by rumpl**
**- What I did**

Invoke containerd's `AppendInfoHandlerWrapper` to add CRI-compatible annotations in the form of labels to pulled content, in order to enable remote snapshotters such as nydus, stargz, and more.

Upstreams:
* https://github.com/rumpl/moby/pull/42

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
cc @ktock

File: vendor.mod
Comment (AkihiroSuda) [inline vendor.mod:32]: Can we use the latest version

Comment (rumpl) [reply vendor.mod:32]: The highest we can go is [v0.11.4](https://github.com/containerd/stargz-snapshotter/releases/tag/v0.11.4), this is the last version that contains `github.com/containerd/stargz-snapshotter/snapshot/overlayutils` used by the version of buildkit that we have, so we need to update buildkit in order to update the stargz snapshotter.

Comment (ktock) [reply vendor.mod]: @AkihiroSuda @rumpl I think v0.11.4 is fine now because this patch uses only `fs/source` pkg and there is no difference between v0.11.4 and the latest in that pkg. In the future, when we import other part of stargz-snapshotter code into moby, we might need to upgrade it to the latest.

**Comment by ktock** [comment]
Thanks for this work! 
LGTM

**Review by AkihiroSuda — DISMISSED** [review]

**Review by vvoland — DISMISSED** [review]

**Comment by imeoer** [comment]
Thanks for the work! `nydus-snapshotter` can be upgraded to `v0.5.0`, others LGTM.

File: vendor.mod
Comment (imeoer) [inline vendor.mod:31]: Better to upgrade to `v0.5.0`.

**Review by imeoer — APPROVED** [review]

**Comment by tianon** [comment]
If we were to instead support a user-specified containerd snapshotter, wouldn't we get similar benefits but without pulling in additional dependencies / code surface we have to maintain over time?

In other words, what specific benefits do we gain by _explicitly_ supporting (and thus maintaining over time) these directly instead of supporting them implicitly via containerd's existing support?

File: vendor.mod
Comment (thaJeztah) [reply vendor.mod]: > The highest we can go is [v0.11.4](https://github.com/containerd/stargz-snapshotter/releases/tag/v0.11.4), this is the last version that contains github.com/containerd/stargz-snapshotter/snapshot/overlayutils used by the version of buildkit that we have, so we need to update buildkit in order to update the stargz snapshotter.

Related PR for 0.10 is in https://github.com/moby/buildkit/pull/3318. It's merged, but there's no new 0.10.x release yet

perhaps @tonistiigi knows if there's plans to do more 0.10 patch releases (for vendoring / go modules)

**Comment by thaJeztah** [comment]
> In other words, what specific benefits do we gain by explicitly supporting (and thus maintaining over time) these directly instead of supporting them implicitly via containerd's existing support?

Also curious about that part; if possible it would be good that we wouldn't have to make changes for every possible snapshotter out there. Not sure what exactly is needed on our side for that. Otherwise we'd continue to play catch-up every time a new snapshotter comes to existence/

**Comment by rumpl** [comment]
> > In other words, what specific benefits do we gain by explicitly supporting (and thus maintaining over time) these directly instead of supporting them implicitly via containerd's existing support?
> 
> Also curious about that part; if possible it would be good that we wouldn't have to make changes for every possible snapshotter out there. Not sure what exactly is needed on our side for that. Otherwise we'd continue to play catch-up every time a new snapshotter comes to existence/

The lazy snapshotters need the image handlers, they add metadata so that the snapshotter knows later what’s what and be able to pull the layers on demand. 

A user-defined snapshotter will only work if that snapshotter doesn’t need any custom image handlers on pull

I only added the snapshotters that are subprojects of the containerd project, I don’t think there will be new ones popping up every week.

**Comment by rumpl** [comment]
> In other words, what specific benefits do we gain by _explicitly_ supporting (and thus maintaining over time) these directly instead of supporting them implicitly via containerd's existing support?

To rephrase my other comment: without this code these two snapshotters don’t work at all, containerd’s client doesn’t support them by default. Maybe this will change in the future with the Transport service they are making but I’m not sure

**Comment by tianon** [comment]
> I only added the snapshotters that are subprojects of the containerd project, I don’t think there will be new ones popping up every week.

Right, but the bar for becoming a non-core project is fairly low, so even if it doesn't happen often, it also doesn't really tell us anything about the support or longevity of the project (and thus whether it makes sense for us to commit to supporting; see the saga around several of our log drivers for a painful example of this 😔).

**Comment by AkihiroSuda** [comment]
The snapshotter-specific code just appends labels, so it is easier to maintain compared to logging drivers.

If this is still a burden to get the PR merged, I guess we can move the snapshot-specific labeling part to `daemon.json` values.

**Comment by thaJeztah** [comment]
> The snapshotter-specific code just appends labels, so it is easier to maintain compared to logging drivers.
>
> If this is still a burden to get the PR merged, I guess we can move the snapshot-specific labeling part to daemon.json values.

If these are well-known (fixed) labels, perhaps they can be set without having to introduce the dependencies (see https://github.com/docker/buildx/pull/1549#discussion_r1085582140 for a similar discussion). Configurable through `daemon.json` (or other ways) _could_ be an option yes (worth exploring).

FWIW; if we only need the dependency for the labels, I'm not so much concerned about the dependency itself (which would then be _mostly_ "dead code" (still not ideal)), but also the ripple effect those dependencies may cause (indirect dependencies, versions of those), which becomes more problematic if we're an actual go module.

**Comment by rumpl** [comment]
> > The snapshotter-specific code just appends labels, so it is easier to maintain compared to logging drivers.
> > If this is still a burden to get the PR merged, I guess we can move the snapshot-specific labeling part to daemon.json values.
> 
> If these are well-known (fixed) labels, perhaps they can be set without having to introduce the dependencies (see [docker/buildx#1549 (comment)](https://github.com/docker/buildx/pull/1549#discussion_r1085582140) for a similar discussion). Configurable through `daemon.json` (or other ways) _could_ be an option yes (worth exploring).

The labels these two snapshotters set don't seem to be well-known, I'm not sure we want to implement a whole logic around adding different labels to different media-types with specific values. This would _really_ mean we are maintaining lazy snapshotters but in a completely different way the original snapshotter does.

> 
> FWIW; if we only need the dependency for the labels, I'm not so much concerned about the dependency itself (which would then be _mostly_ "dead code" (still not ideal)), but also the ripple effect those dependencies may cause (indirect dependencies, versions of those), which becomes more problematic if we're an actual go module.

FYI we already had estargz as a dependency because buildkit handles it and we will also depend on nydus once we [update buildkit](https://github.com/moby/buildkit/blob/master/go.mod#L23)

**Comment by neersighted** [comment]
It was observed today in the maintainer's call that containerd itself does not import these snapshotters, and supports them just fine. It appears that more-or-less the same code being imported and called here is built into the CRI:

https://github.com/containerd/containerd/blob/dadd203c255756d2208879088f7c61e1bee7d1f8/pkg/cri/sbserver/image_pull.go#L577-L602

Can we see if we can incorporate this code directly, and raise a code against containerd that hoists it out of the CRI and to an exported package? Then we can add a FIXME to use the newly exported method when we eventually pick up the change in containerd.

**Comment by ktock** [comment]
@neersighted Thank you for the suggestion! Yes, using [`appendInfoHandlerWrapper`](https://github.com/containerd/containerd/blob/dadd203c255756d2208879088f7c61e1bee7d1f8/pkg/cri/sbserver/image_pull.go#L556-L624) with [`WithImageHandlerWrapper`](https://github.com/containerd/containerd/blob/v1.7.0-beta.3/client_opts.go#L227) is enough to integrate remote snapshotters aware of these labels (e.g. stargz-snapshotter and maybe nydus as well cc: @imeoer), without importing each project's handler.

As of now, `appendInfoHandlerWrapper` isn't an exported function so we can work on it in the following steps:

1. This PR (cc @rumpl): Copy [`appendInfoHandlerWrapper`](https://github.com/containerd/containerd/blob/dadd203c255756d2208879088f7c61e1bee7d1f8/pkg/cri/sbserver/image_pull.go#L556-L624) to moby's codebase and use it. Add FIXME comment to use the function exported by containerd in the following-up patch.
2. Following-up PR at containerd project: Fix containerd to export that function (Under discussion on https://github.com/containerd/containerd/pull/8036)
3. Following-up PR at moby project: Fix moby's codebase to use the newly exported function when the version of containerd including the above fix is imported to moby.

I'm willing to working on following-up PRs.

cc @tianon @thaJeztah @AkihiroSuda

**Comment by imeoer** [comment]
Thanks @ktock for the work, we will update nydus snapshotter once https://github.com/containerd/containerd/pull/8036 is merged and released.

**Review by neersighted — CHANGES_REQUESTED** [review]
Adding my review as a blocker until the new "don't import the snapshotters directly" approach can be implemented.

**Comment by imeoer** [comment]
[nydus snapshotter](https://github.com/containerd/nydus-snapshotter/pull/351) has integrated https://github.com/containerd/containerd/pull/8036
cc @rumpl Any progress we can push forward together?

**Review by neersighted — DISMISSED** [review]
This can be rebased in terms of #44982 once that PR is merged; we can also bump to containerd 1.7 if desired after that PR is merged.

**Comment by neersighted** [comment]
I've gone ahead and rebased this in terms of containerd 1.6.18 and the newly exported code; PTAL @vvoland @rumpl

**Comment by rumpl** [comment]
Can't approve my own PR but 👍 LGTM

**Review by AkihiroSuda — APPROVED** [review]

**Review by ktock — APPROVED** [review]

**Comment by neersighted** [comment]
Bringing this in; and I'm really happy to see how we came together across projects to improve this for the ecosystem :tada:

**Pull request merged**
